[homekit] IllegalStateException: Duplicate key Temperature_GF_Kitchen

When you install the Demo package the Homekit add-on throws the following exception after installation/startup:

```
[ERROR] [org.openhab.io.homekit               ] - bundle org.openhab.io.homekit:2.5.0.201905160427 (241)[org.openhab.io.homekit.internal.HomekitImpl(233)] : The setItemRegistry method has thrown an exception
java.lang.IllegalStateException: Duplicate key Temperature_GF_Kitchen (Type=NumberItem, State=NULL, Label=Kitchen, Category=temperature, Tags=[Measurement, Temperature, CurrentTemperature], Groups=[Temperature, GF_Kitchen, Thermostat])
	at java.util.stream.Collectors.lambda$throwingMerger$0(Collectors.java:133) ~[?:?]
	at java.util.HashMap.merge(HashMap.java:1254) ~[?:?]
	at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1320) ~[?:?]
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169) ~[?:?]
	at java.util.Iterator.forEachRemaining(Iterator.java:116) ~[?:?]
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[?:?]
	at org.openhab.io.homekit.internal.accessories.HomekitAccessoryFactory.getCharacteristicItems(HomekitAccessoryFactory.java:147) ~[?:?]
	at org.openhab.io.homekit.internal.accessories.HomekitAccessoryFactory.create(HomekitAccessoryFactory.java:49) ~[?:?]
	at org.openhab.io.homekit.internal.HomekitChangeListener.createRootAccessory(HomekitChangeListener.java:184) ~[?:?]
	at org.openhab.io.homekit.internal.HomekitChangeListener.lambda$7(HomekitChangeListener.java:156) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[?:?]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) ~[?:?]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) ~[?:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:?]
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1556) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) ~[?:?]
	at org.openhab.io.homekit.internal.HomekitChangeListener.maybeInitialize(HomekitChangeListener.java:156) ~[?:?]
	at org.openhab.io.homekit.internal.HomekitChangeListener.setItemRegistry(HomekitChangeListener.java:140) ~[?:?]
	at org.openhab.io.homekit.internal.HomekitImpl.setItemRegistry(HomekitImpl.java:66) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:228) ~[42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41) ~[42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:664) ~[42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:510) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.inject.methods.BindMethod.invoke(BindMethod.java:42) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.DependencyManager.doInvokeBindMethod(DependencyManager.java:1813) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.DependencyManager.bindDependency(DependencyManager.java:1637) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.DependencyManager.bind(DependencyManager.java:1624) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:307) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:114) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:982) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:955) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:900) [42:org.apache.felix.scr:2.1.16]
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse$1.run(ServiceFactoryUse.java:212) [?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:210) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.getService(ServiceFactoryUse.java:111) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceConsumer$2.getService(ServiceConsumer.java:45) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:508) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:461) [?:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:624) [?:?]
	at com.eclipsesource.jaxrs.publisher.internal.ResourceTracker.addingService(ResourceTracker.java:39) [20:com.eclipsesource.jaxrs.publisher:5.3.1.201602281253]
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:941) [?:?]
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:870) [?:?]
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256) [?:?]
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229) [?:?]
	at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:901) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109) [?:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:920) [?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) [?:?]
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:127) [?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:225) [?:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:469) [?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:906) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:892) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.RegistrationManager.changeRegistration(RegistrationManager.java:128) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.registerService(AbstractComponentManager.java:959) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:732) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:666) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:432) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:665) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:338) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:382) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.Activator.access$200(Activator.java:49) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:264) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169) [42:org.apache.felix.scr:2.1.16]
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49) [42:org.apache.felix.scr:2.1.16]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:482) [?:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:415) [?:?]
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232) [?:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:444) [?:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:908) [?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) [?:?]
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:213) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:120) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:112) [?:?]
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:168) [?:?]
	at org.eclipse.osgi.container.Module.publishEvent(Module.java:476) [?:?]
	at org.eclipse.osgi.container.Module.start(Module.java:467) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1634) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1614) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1585) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1528) [?:?]
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1) [?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) [?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340) [?:?]
```

Perhaps it's caused by some of the changes in #5581?

--- a/bundles/org.openhab.io.homekit/README.md
+++ b/bundles/org.openhab.io.homekit/README.md
@@ -64,17 +64,18 @@ Complex accessories require a tag on a Group indicating the accessory type, as w
 
 A full list of supported accessory types can be found in the table below.
 
-| Tag                   | Child tag                     | Supported items           | Description                                                                                                                                                                                                                                   |
-|--------------------   |----------------------------   |-----------------------    |---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------  |
-| Lighting              |                               | Switch, Dimmer, Color     | A lightbulb, switchable, dimmable or rgb                                                                                                                                                                                                      |
-| Switchable            |                               | Switch, Dimmer, Color     | An accessory that can be turned off and on. While similar to a lightbulb, this will be presented differently in the Siri grammar and iOS apps                                                                                                 |
-| ContactSensor         |                               | Contact                   | An accessory with on/off state that can be viewed in HomeKit but not changed such as a contact sensor for a door or window                                                                                                                    |
-| CurrentTemperature    |                               | Number                    | An accessory that provides a single read-only temperature value. The units default to celsius but can be overridden globally using the useFahrenheitTemperature global property                                                               |
-| CurrentHumidity       |                               | Number                    | An accessory that provides a single read-only value indicating the relative humidity.                                                                                                                                                         |
-| Thermostat            |                               | Group                     | A thermostat requires all child tags defined below                                                                                                                                                                                            |
-|                       | CurrentTemperature            | Number                    | The current temperature, same as above                                                                                                                                                                                                        |
-|                       | homekit:HeatingCoolingMode    | String                    | Indicates the current mode of the device: OFF, AUTO, HEAT, COOL. The string's value must match those defined in the thermostat*Mode properties. This is a HomeKit-specific term and therefore the tags needs to be prefixed with "homekit:"   |
-|                       | TargetTemperature             | Number                    | A target temperature that will engage the thermostat's heating and cooling actions as necessary, depending on the heatingCoolingMode                                                                                                          |
+| Tag                   | Child tag                         | Supported items           | Description                                                                                                                                                                                                                                   |
+|--------------------   |--------------------------------   |-----------------------    |---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------  |
+| Lighting              |                                   | Switch, Dimmer, Color     | A lightbulb, switchable, dimmable or rgb                                                                                                                                                                                                      |
+| Switchable            |                                   | Switch, Dimmer, Color     | An accessory that can be turned off and on. While similar to a lightbulb, this will be presented differently in the Siri grammar and iOS apps                                                                                                 |
+| ContactSensor         |                                   | Contact                   | An accessory with on/off state that can be viewed in HomeKit but not changed such as a contact sensor for a door or window                                                                                                                    |
+| CurrentTemperature    |                                   | Number                    | An accessory that provides a single read-only temperature value. The units default to celsius but can be overridden globally using the useFahrenheitTemperature global property                                                               |
+| CurrentHumidity       |                                   | Number                    | An accessory that provides a single read-only value indicating the relative humidity.                                                                                                                                                         |
+| Thermostat            |                                   | Group                     | A thermostat requires all child tags defined below                                                                                                                                                                                            |
+|                       | CurrentTemperature                | Number                    | The current temperature, same as above                                                                                                                                                                                                        |
+|                       | homekit:TargetHeatingCoolingMode  | String                    | Indicates the desired mode of the device: OFF, AUTO, HEAT, COOL. The string's value must match those defined in the thermostat*Mode properties. This is a HomeKit-specific term and therefore the tags needs to be prefixed with "homekit:"   |
+|                       | homekit:CurrentHeatingCoolingMode | String                    | Indicates the current mode of the device: OFF, AUTO, HEAT, COOL. The string's value must match those defined in the thermostat*Mode properties. This is a HomeKit-specific term and therefore the tags needs to be prefixed with "homekit:"   |
+|                       | homekit:TargetTemperature         | Number                    | A target temperature that will engage the thermostat's heating and cooling actions as necessary, depending on the heatingCoolingMode. This is a HomeKit-specific term and therefore the tags needs to be prefixed with "homekit:"             |
 
 See the sample below for example items:
 
@@ -85,7 +86,7 @@ Number BedroomTemperature "Bedroom Temperature" (gBedroom) [ "CurrentTemperature
 Group gDownstairsThermostat "Downstairs Thermostat" (gFF) [ "Thermostat" ]
 Number DownstairsThermostatCurrentTemp "Downstairs Thermostat Current Temperature" (gDownstairsThermostat) [ "CurrentTemperature" ]
 Number DownstairsThermostatTargetTemperature "Downstairs Thermostat Target Temperature" (gDownstairsThermostat) [ "TargetTemperature" ]
-String DownstairsThermostatHeatingCoolingMode "Downstairs Thermostat Heating/Cooling Mode" (gDownstairsThermostat) [ "homekit:HeatingCoolingMode" ]
+String DownstairsThermostatHeatingCoolingMode "Downstairs Thermostat Heating/Cooling Mode" (gDownstairsThermostat) [ "homekit:TargetHeatingCoolingMode" ]
 ```
 
 ## Common Problems
@@ -109,7 +110,7 @@ This unique identifier is hashed from the Item's name.
 For that reason, it is important that the name of your Items exposed to HomeKit remain consistent.
 
 HomeKit listens by default on port 9124.
-Java perfers the IPv6 network stack by default.
+Java prefers the IPv6 network stack by default.
 If you have connection or detection problems, you can configure Java to prefer the IPv4 network stack instead.
 To prefer the IPv4 network stack, adapt the Java command line arguments to include: `-Djava.net.preferIPv4Stack=true`
 Depending on the openHAB installation method, you should modify `start.sh`, `start_debug.sh`, `start.bat`, or `start_debug.bat` (standalone/manual installation) or `EXTRA_JAVA_OPTS` in `/etc/default/openhab2` (Debian installation).

--- a/bundles/org.openhab.io.homekit/README.md
+++ b/bundles/org.openhab.io.homekit/README.md
@@ -7,6 +7,8 @@ In order to do so, you will need to make some configuration changes.
 HomeKit organizes your home into "accessories" that are made up of a number of "characteristics".
 Some accessory types require a specific set of characteristics.
 
+**Attention: Some tags have been renamed. Old style may not be supported in future versions. See below for details.**
+
 ## Global Configuration
 
 Your first step will be to create the `homekit.cfg` in your `$OPENHAB_CONF/services` folder.
@@ -76,6 +78,8 @@ A full list of supported accessory types can be found in the table below.
 |                       | homekit:TargetHeatingCoolingMode  | String                    | Indicates the desired mode of the device: OFF, AUTO, HEAT, COOL. The string's value must match those defined in the thermostat*Mode properties. This is a HomeKit-specific term and therefore the tags needs to be prefixed with "homekit:"   |
 |                       | homekit:CurrentHeatingCoolingMode | String                    | Indicates the current mode of the device: OFF, AUTO, HEAT, COOL. The string's value must match those defined in the thermostat*Mode properties. This is a HomeKit-specific term and therefore the tags needs to be prefixed with "homekit:"   |
 |                       | homekit:TargetTemperature         | Number                    | A target temperature that will engage the thermostat's heating and cooling actions as necessary, depending on the heatingCoolingMode. This is a HomeKit-specific term and therefore the tags needs to be prefixed with "homekit:"             |
+
+**Please note:** `TargetTemperature` has been renamed to `homekit:TagretTemperature` and `homekit:HeatingCoolingMode` has been renamed to `homekit:TargetHeatingCoolingMode`.
 
 See the sample below for example items:
 

--- a/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitAccessoryType.java
+++ b/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitAccessoryType.java
@@ -14,6 +14,12 @@ package org.openhab.io.homekit.internal;
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import org.eclipse.smarthome.core.items.Item;
+import org.eclipse.smarthome.core.library.items.ColorItem;
+import org.eclipse.smarthome.core.library.items.DimmerItem;
 
 /**
  * Enum of the possible device types. The defined tag string can be used
@@ -58,7 +64,32 @@ public enum HomekitAccessoryType {
         return tag;
     }
 
+    /**
+     * get accessoryType from String
+     *
+     * @param tag the tag string
+     * @return accessoryType or null if not found
+     */
     public static HomekitAccessoryType valueOfTag(String tag) {
         return TAG_MAP.get(tag);
+    }
+
+    /**
+     * get accessoryType for a given Item
+     *
+     * @param item the item
+     * @return accessoryType or null if not found
+     */
+    public static HomekitAccessoryType fromItem(Item item) {
+        Set<String> tags = item.getTags();
+        HomekitAccessoryType accessoryType = tags.stream().map(tag -> TAG_MAP.get(tag)).filter(Objects::nonNull)
+                .findFirst().orElse(null);
+        if (item instanceof ColorItem && accessoryType == LIGHTBULB) {
+            return COLORFUL_LIGHTBULB;
+        } else if (item instanceof DimmerItem && accessoryType == LIGHTBULB) {
+            return DIMMABLE_LIGHTBULB;
+        } else {
+            return accessoryType;
+        }
     }
 }

--- a/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitChangeListener.java
+++ b/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitChangeListener.java
@@ -176,15 +176,15 @@ public class HomekitChangeListener implements ItemRegistryChangeListener {
         try {
             if (taggedItem.isMemberOfAccessoryGroup()) {
                 logger.warn("Bug! Cannot add {} as a root accessory if it is a member of a group! ",
-                        taggedItem.getName());
+                        taggedItem.getItem().getUID());
                 return;
             }
-            logger.debug("Adding homekit device {}", taggedItem.getItem().getName());
+            logger.debug("Adding homekit device {}", taggedItem.getItem().getUID());
             accessoryRegistry.addRootAccessory(taggedItem.getName(),
                     HomekitAccessoryFactory.create(taggedItem, itemRegistry, updater, settings));
-            logger.debug("Added homekit device {}", taggedItem.getItem().getName());
+            logger.debug("Added homekit device {}", taggedItem.getItem().getUID());
         } catch (HomekitException | IncompleteAccessoryException e) {
-            logger.warn("Could not add device: {}", e.getMessage(), e);
+            logger.warn("Could not add device {}: {}", taggedItem.getItem().getUID(), e.getMessage());
         }
     }
 }

--- a/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitCharacteristicType.java
+++ b/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitCharacteristicType.java
@@ -14,6 +14,9 @@ package org.openhab.io.homekit.internal;
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
+
+import org.eclipse.smarthome.core.items.Item;
 
 /**
  * Characteristics are used by complex accessories that can't be represented by
@@ -53,7 +56,23 @@ public enum HomekitCharacteristicType {
         return tag;
     }
 
+    /**
+     * get characteristicType from String
+     *
+     * @param tag the tag string
+     * @return characteristicType or null if not found
+     */
     public static HomekitCharacteristicType valueOfTag(String tag) {
         return TAG_MAP.get(tag);
+    }
+
+    /**
+     * get characteristicType for a given Item
+     *
+     * @param item the item
+     * @return characteristicType or null if not found
+     */
+    public static HomekitCharacteristicType fromItem(Item item) {
+        return item.getTags().stream().map(tag -> TAG_MAP.get(tag)).filter(Objects::nonNull).findFirst().orElse(null);
     }
 }

--- a/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitTaggedItem.java
+++ b/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitTaggedItem.java
@@ -14,7 +14,6 @@ package org.openhab.io.homekit.internal;
 
 import java.util.List;
 import java.util.Map;
-import java.util.Objects;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -23,8 +22,6 @@ import org.apache.commons.lang.builder.HashCodeBuilder;
 import org.eclipse.smarthome.core.items.GroupItem;
 import org.eclipse.smarthome.core.items.Item;
 import org.eclipse.smarthome.core.items.ItemRegistry;
-import org.eclipse.smarthome.core.library.items.ColorItem;
-import org.eclipse.smarthome.core.library.items.DimmerItem;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -55,32 +52,12 @@ public class HomekitTaggedItem {
     private final int id;
     private GroupItem parentGroupItem;
 
-    private static Stream<String> getPrefixedTags(Item item) {
-        Stream<String> prefixedTags = item.getTags().stream();
-        if (item instanceof ColorItem) {
-            prefixedTags = prefixedTags.map(tag -> "Colorful" + tag);
-        } else if (item instanceof DimmerItem) {
-            prefixedTags = prefixedTags.map(tag -> "Dimmable" + tag);
-        }
-        return prefixedTags;
-    }
-
-    public static HomekitAccessoryType findAccessoryType(Item item) {
-        return getPrefixedTags(item).map(tag -> HomekitAccessoryType.valueOfTag(tag)).filter(Objects::nonNull)
-                .findFirst().orElseGet(() -> null);
-    }
-
-    public static HomekitCharacteristicType findCharacteristicType(Item item) {
-        return getPrefixedTags(item).map(tag -> HomekitCharacteristicType.valueOfTag(tag)).filter(Objects::nonNull)
-                .findFirst().orElseGet(() -> null);
-    }
-
     public HomekitTaggedItem(Item item, ItemRegistry itemRegistry) {
         this.item = item;
 
         try {
-            homekitAccessoryType = findAccessoryType(item);
-            homekitCharacteristicType = findCharacteristicType(item);
+            homekitAccessoryType = HomekitAccessoryType.fromItem(item);
+            homekitCharacteristicType = HomekitCharacteristicType.fromItem(item);
             if (homekitAccessoryType != null && homekitCharacteristicType != null) {
                 throw new BadItemConfigurationException(
                         "Items cannot be tagged as both a characteristic and an accessory type");
@@ -205,7 +182,7 @@ public class HomekitTaggedItem {
         return id;
     }
 
-    static List<GroupItem> findMyAccessoryGroups(Item item, ItemRegistry itemRegistry) {
+    public static List<GroupItem> findMyAccessoryGroups(Item item, ItemRegistry itemRegistry) {
         return item.getGroupNames().stream().flatMap(name -> {
             Item groupItem = itemRegistry.get(name);
             if ((groupItem != null) && (groupItem instanceof GroupItem)) {

--- a/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java
+++ b/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java
@@ -13,9 +13,9 @@
 package org.openhab.io.homekit.internal.accessories;
 
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
-import java.util.stream.Collectors;
 
 import org.eclipse.smarthome.core.items.GroupItem;
 import org.eclipse.smarthome.core.items.Item;
@@ -143,9 +143,19 @@ public class HomekitAccessoryFactory {
     private static Map<HomekitCharacteristicType, Item> getCharacteristicItems(HomekitTaggedItem taggedItem) {
         if (taggedItem.isGroup()) {
             GroupItem groupItem = (GroupItem) taggedItem.getItem();
-            Map<HomekitCharacteristicType, Item> characteristicItems = groupItem.getMembers().stream()
-                    .collect(Collectors.toMap(item -> HomekitTaggedItem.findCharacteristicType(item), item -> item));
-            characteristicItems.entrySet().removeIf(e -> e.getKey() == null);
+            Map<HomekitCharacteristicType, Item> characteristicItems = new HashMap<>();
+            groupItem.getMembers().forEach(item -> {
+                HomekitCharacteristicType type = HomekitCharacteristicType.fromItem(item);
+                if (type != null) {
+                    if (characteristicItems.containsKey(type)) {
+                        LOGGER.warn("incorrect configuration for {} detected: {} and {} are tagged as {}, skipping {}",
+                                taggedItem.getItem().getUID(), characteristicItems.get(type).getUID(), item.getUID(),
+                                type, item.getUID());
+                    } else {
+                        characteristicItems.put(type, item);
+                    }
+                }
+            });
             return Collections.unmodifiableMap(characteristicItems);
         } else {
             // do nothing; only accessory groups have characteristic items

--- a/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitThermostatImpl.java
+++ b/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitThermostatImpl.java
@@ -19,6 +19,7 @@ import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 
 import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.smarthome.core.items.GroupItem;
 import org.eclipse.smarthome.core.items.Item;
 import org.eclipse.smarthome.core.items.ItemRegistry;
@@ -41,28 +42,21 @@ import com.beowulfe.hap.accessories.thermostat.BasicThermostat;
 /**
  * Implements Thermostat as a GroupedAccessory made up of multiple items:
  * <ul>
- * <li>Cooling Threshold: Decimal type</li>
- * <li>Heating Threshold: Decimal type</li>
- * <li>Auto Threshold: Decimal type</li>
- * <li>Current Temperature: Decimal type</li>
- * <li>Heating/Cooling Mode: String type (see HomekitSettings.thermostat*Mode)</li>
+ * <li>Current Temperature: Number type</li>
+ * <li>Target Temperature: Number type</li>
+ * <li>Current Heating/Cooling Mode: String type (see HomekitSettings.thermostat*Mode)</li>
+ * <li>Target Heating/Cooling Mode: String type (see HomekitSettings.thermostat*Mode)</li>
  * </ul>
  *
  * @author Andy Lintner - Initial contribution
  */
 class HomekitThermostatImpl extends AbstractTemperatureHomekitAccessoryImpl<GroupItem> implements BasicThermostat {
-
     private final HomekitSettings settings;
-    @NonNull
-    private NumberItem currentTemperatureItem;
-    @NonNull
-    private StringItem targetHeatingCoolingModeItem;
 
-    @NonNull
-    private Optional<StringItem> heatingCoolingCurrentModeItem;
-
-    @NonNull
-    private NumberItem targetTemperatureItem;
+    private @NonNull NumberItem currentTemperatureItem;
+    private @NonNull StringItem targetHeatingCoolingModeItem;
+    private @Nullable StringItem currentHeatingCoolingModeItem;
+    private @NonNull NumberItem targetTemperatureItem;
 
     private Logger logger = LoggerFactory.getLogger(HomekitThermostatImpl.class);
 
@@ -74,21 +68,44 @@ class HomekitThermostatImpl extends AbstractTemperatureHomekitAccessoryImpl<Grou
 
         HashMap<HomekitCharacteristicType, Item> characteristicItems = new HashMap<>(origCharacteristicItems);
         this.settings = settings;
-        this.currentTemperatureItem = (NumberItem) currentTemperatureItem;
 
-        this.targetHeatingCoolingModeItem = getItemWithDeprecation(characteristicItems,
+        if (currentTemperatureItem instanceof NumberItem) {
+            this.currentTemperatureItem = (NumberItem) currentTemperatureItem;
+        } else {
+            throw new IncompleteAccessoryException(currentTemperatureItem.getUID()
+                    + "tagged as thermostat currentTemperatureItem has wrong item type (NumberItem needed)");
+        }
+
+        Item targetHeatingCoolingModeItem = getItemWithDeprecation(characteristicItems,
                 HomekitCharacteristicType.TARGET_HEATING_COOLING_MODE,
-                HomekitCharacteristicType.OLD_TARGET_HEATING_COOLING_MODE).map(m -> (StringItem) m).orElseThrow(
+                HomekitCharacteristicType.OLD_TARGET_HEATING_COOLING_MODE).orElseThrow(
                         () -> new IncompleteAccessoryException(HomekitCharacteristicType.TARGET_HEATING_COOLING_MODE));
+        if (targetHeatingCoolingModeItem instanceof StringItem) {
+            this.targetHeatingCoolingModeItem = (StringItem) targetHeatingCoolingModeItem;
+        } else {
+            throw new IncompleteAccessoryException(targetHeatingCoolingModeItem.getUID()
+                    + " tagged as thermostat targetHeatingCoolingMode has wrong item type (String needed)");
+        }
 
-        this.targetTemperatureItem = getItemWithDeprecation(characteristicItems,
+        Item targetTemperatureItem = getItemWithDeprecation(characteristicItems,
                 HomekitCharacteristicType.TARGET_TEMPERATURE, HomekitCharacteristicType.OLD_TARGET_TEMPERATURE)
-                        .map(item -> (NumberItem) item).orElseThrow(
+                        .orElseThrow(
                                 () -> new IncompleteAccessoryException(HomekitCharacteristicType.TARGET_TEMPERATURE));
+        if (targetTemperatureItem instanceof NumberItem) {
+            this.targetTemperatureItem = (NumberItem) targetTemperatureItem;
+        } else {
+            throw new IncompleteAccessoryException(targetTemperatureItem.getUID()
+                    + " tagged as thermostat targetTemperature has wrong item type (Number needed)");
+        }
 
-        this.heatingCoolingCurrentModeItem = Optional
-                .ofNullable(characteristicItems.remove(HomekitCharacteristicType.CURRENT_HEATING_COOLING_STATE))
-                .map(m -> (StringItem) m);
+        Item currentHeatingCoolingModeItem = characteristicItems
+                .remove(HomekitCharacteristicType.CURRENT_HEATING_COOLING_STATE);
+        if (currentHeatingCoolingModeItem instanceof StringItem || currentHeatingCoolingModeItem == null) {
+            this.currentHeatingCoolingModeItem = (StringItem) currentHeatingCoolingModeItem;
+        } else {
+            throw new IncompleteAccessoryException(currentHeatingCoolingModeItem.getUID()
+                    + " tagged as thermostat currentHeatingCoolingMode has wrong item type (String needed)");
+        }
 
         characteristicItems.entrySet().stream().forEach(entry -> {
             logger.warn("Item {} has unrecognized thermostat characteristic: {}", entry.getValue().getName(),
@@ -111,8 +128,10 @@ class HomekitThermostatImpl extends AbstractTemperatureHomekitAccessoryImpl<Grou
 
     @Override
     public CompletableFuture<ThermostatMode> getCurrentMode() {
-        String stringValue = heatingCoolingCurrentModeItem.map(i -> i.getState().toString())
-                .orElseGet(() -> settings.getCurrentModeOff());
+        String stringValue = settings.getCurrentModeOff();
+        if (currentHeatingCoolingModeItem != null) {
+            stringValue = currentHeatingCoolingModeItem.getState().toString();
+        }
         ThermostatMode mode;
 
         if (stringValue.equalsIgnoreCase(settings.getThermostatCurrentModeCooling())) {
@@ -207,7 +226,9 @@ class HomekitThermostatImpl extends AbstractTemperatureHomekitAccessoryImpl<Grou
 
     @Override
     public void subscribeCurrentMode(HomekitCharacteristicChangeCallback callback) {
-        heatingCoolingCurrentModeItem.ifPresent(item -> getUpdater().subscribe(item, callback));
+        if (currentHeatingCoolingModeItem != null) {
+            getUpdater().subscribe(currentHeatingCoolingModeItem, callback);
+        }
     }
 
     @Override

--- a/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitThermostatImpl.java
+++ b/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitThermostatImpl.java
@@ -108,7 +108,7 @@ class HomekitThermostatImpl extends AbstractTemperatureHomekitAccessoryImpl<Grou
         }
 
         characteristicItems.entrySet().stream().forEach(entry -> {
-            logger.warn("Item {} has unrecognized thermostat characteristic: {}", entry.getValue().getName(),
+            logger.warn("Item {} has unrecognized thermostat characteristic: {}", entry.getValue().getUID(),
                     entry.getKey().getTag());
         });
     }
@@ -120,7 +120,7 @@ class HomekitThermostatImpl extends AbstractTemperatureHomekitAccessoryImpl<Grou
             targetTempItem = Optional.ofNullable(characteristicItems.remove(deprecatedTag));
             targetTempItem.ifPresent(item -> {
                 logger.warn("The tag {} has been renamed to {}; please update your things, accordingly",
-                        currentTag.getTag(), deprecatedTag.getTag());
+                        deprecatedTag.getTag(), currentTag.getTag());
             });
         }
         return targetTempItem;

--- a/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/IncompleteAccessoryException.java
+++ b/bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/IncompleteAccessoryException.java
@@ -20,10 +20,12 @@ import org.openhab.io.homekit.internal.HomekitCharacteristicType;
  */
 public class IncompleteAccessoryException extends Exception {
     private static final long serialVersionUID = 8595808359805444177L;
-    final HomekitCharacteristicType missingType;
 
     public IncompleteAccessoryException(HomekitCharacteristicType missingType) {
         super(String.format("Missing accessory type %s", missingType.getTag()));
-        this.missingType = missingType;
+    }
+
+    public IncompleteAccessoryException(String message) {
+        super(message);
     }
 }

